File: clients/js/src/generated-maiker/instructions/claimFee.ts
import { TransactionInstruction, PublicKey, AccountMeta } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@coral-xyz/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface ClaimFeeAccounts {
  authority: PublicKey
  globalConfig: PublicKey
  strategy: PublicKey
  strategyVaultX: PublicKey
  strategyVaultY: PublicKey
  position: PublicKey
  lbPair: PublicKey
  binArrayLower: PublicKey
  binArrayUpper: PublicKey
  reserveX: PublicKey
  reserveY: PublicKey
  tokenXMint: PublicKey
  tokenYMint: PublicKey
  /** The lb_clmm program */
  lbClmmProgram: PublicKey
  eventAuthority: PublicKey
  tokenProgram: PublicKey
}

export function claimFee(
  accounts: ClaimFeeAccounts,
  programId: PublicKey = PROGRAM_ID
) {
  const keys: Array<AccountMeta> = [
    { pubkey: accounts.authority, isSigner: true, isWritable: false },
    { pubkey: accounts.globalConfig, isSigner: false, isWritable: false },
    { pubkey: accounts.strategy, isSigner: false, isWritable: true },
    { pubkey: accounts.strategyVaultX, isSigner: false, isWritable: true },
    { pubkey: accounts.strategyVaultY, isSigner: false, isWritable: true },
    { pubkey: accounts.position, isSigner: false, isWritable: true },
    { pubkey: accounts.lbPair, isSigner: false, isWritable: true },
    { pubkey: accounts.binArrayLower, isSigner: false, isWritable: true },
    { pubkey: accounts.binArrayUpper, isSigner: false, isWritable: true },
    { pubkey: accounts.reserveX, isSigner: false, isWritable: true },
    { pubkey: accounts.reserveY, isSigner: false, isWritable: true },
    { pubkey: accounts.tokenXMint, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenYMint, isSigner: false, isWritable: false },
    { pubkey: accounts.lbClmmProgram, isSigner: false, isWritable: false },
    { pubkey: accounts.eventAuthority, isSigner: false, isWritable: false },
    { pubkey: accounts.tokenProgram, isSigner: false, isWritable: false },
  ]
  const identifier = Buffer.from([169, 32, 79, 137, 136, 232, 70, 137])
  const data = identifier
  const ix = new TransactionInstruction({ keys, programId, data })
  return ix
}

File: programs/maiker-contracts/src/instructions/cpi/claim_fee.rs
use crate::state::*;
use anchor_lang::prelude::*;
use anchor_spl::token::{Token, TokenAccount};
use dlmm_interface::{claim_fee_invoke_signed, ClaimFeeAccounts};

#[derive(Accounts)]
pub struct ClaimFee<'info> {
    #[account(
        constraint = authority.key() == global_config.admin
    )]
    pub authority: Signer<'info>,

    #[account(
        seeds = [GlobalConfig::SEED_PREFIX.as_bytes()],
        bump = global_config.bump,
    )]
    pub global_config: Box<Account<'info, GlobalConfig>>,

    #[account(mut)]
    pub strategy: Box<Account<'info, StrategyConfig>>,

    #[account(
        mut,
        token::mint = token_x_mint,
        token::authority = strategy
    )]
    pub strategy_vault_x: Box<Account<'info, TokenAccount>>,

    #[account(
        mut,
        token::mint = token_y_mint,
        token::authority = strategy
    )]
    pub strategy_vault_y: Box<Account<'info, TokenAccount>>,

    /// CHECK: This is the position account
    #[account(mut)]
    pub position: UncheckedAccount<'info>,

    /// CHECK: This is the LB pair account
    #[account(mut)]
    pub lb_pair: UncheckedAccount<'info>,

    /// CHECK: Bin array lower
    #[account(mut)]
    pub bin_array_lower: UncheckedAccount<'info>,

    /// CHECK: Bin array upper
    #[account(mut)]
    pub bin_array_upper: UncheckedAccount<'info>,

    /// CHECK: Reserve account for token X
    #[account(mut)]
    pub reserve_x: UncheckedAccount<'info>,

    /// CHECK: Reserve account for token Y
    #[account(mut)]
    pub reserve_y: UncheckedAccount<'info>,

    /// CHECK: The strategy vault for token X
    pub token_x_mint: UncheckedAccount<'info>,

    /// CHECK: The strategy vault for token Y
    pub token_y_mint: UncheckedAccount<'info>,

    /// The lb_clmm program
    /// CHECK: The lb_clmm program
    #[account(address = dlmm_interface::ID)]
    pub lb_clmm_program: UncheckedAccount<'info>,

    /// CHECK: Event authority for lb_clmm
    pub event_authority: UncheckedAccount<'info>,

    pub token_program: Program<'info, Token>,
}

pub fn claim_fee_handler(ctx: Context<ClaimFee>) -> Result<()> {
    // TODO: Validation Lb Pair and Strategy token accounts

    let strategy_signer = ctx.accounts.strategy.get_pda_signer();
    let strategy_signer_seeds = &[&strategy_signer[..]];

    let accounts = ClaimFeeAccounts {
        lb_pair: &ctx.accounts.lb_pair.to_account_info(),
        position: &ctx.accounts.position.to_account_info(),
        bin_array_lower: &ctx.accounts.bin_array_lower.to_account_info(),
        bin_array_upper: &ctx.accounts.bin_array_upper.to_account_info(),
        sender: &ctx.accounts.strategy.to_account_info(),
        reserve_x: &ctx.accounts.reserve_x.to_account_info(),
        reserve_y: &ctx.accounts.reserve_y.to_account_info(),
        user_token_x: &ctx.accounts.strategy_vault_x.to_account_info(),
        user_token_y: &ctx.accounts.strategy_vault_y.to_account_info(),
        token_x_mint: &ctx.accounts.token_x_mint.to_account_info(),
        token_y_mint: &ctx.accounts.token_y_mint.to_account_info(),
        token_program: &ctx.accounts.token_program.to_account_info(),
        event_authority: &ctx.accounts.event_authority.to_account_info(),
        program: &ctx.accounts.lb_clmm_program.to_account_info(),
    };

    claim_fee_invoke_signed(accounts, strategy_signer_seeds)?;

    Ok(())
}
